File: apps/backend/api/tasks_api.py
"""HTTP API for tasks (user-scoped backlog)."""

from __future__ import annotations

import uuid
from typing import Any, Literal

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

from apps.backend.domain.agent_task_access import (
    user_may_access_task_row,
    user_may_access_workspace,
)
from apps.backend.infrastructure import agent_artifacts_store, agent_tasks_store
from apps.backend.infrastructure.auth import get_current_user
from apps.backend.infrastructure.db import db

router = APIRouter(prefix="/v1/tasks", tags=["tasks"])


class TaskCreateBody(BaseModel):
    scope: Literal["global", "workspace"] = "global"
    goal: str = Field(..., min_length=1, max_length=16000)
    task_type: str = Field(default="general", max_length=128)
    workspace_id: uuid.UUID | None = None
    parent_task_id: uuid.UUID | None = None
    conversation_id: uuid.UUID | None = None
    status: str = Field(default="draft", max_length=32)
    priority: Literal["low", "normal", "high"] = "normal"
    assigned_agent_id: str | None = Field(default=None, max_length=128)
    requirements: list[Any] = Field(default_factory=list)
    artifact_refs: list[str] = Field(default_factory=list)


class TaskPatchBody(BaseModel):
    status: str | None = None
    goal: str | None = Field(default=None, max_length=16000)
    priority: Literal["low", "normal", "high"] | None = None
    assigned_agent_id: str | None = Field(default=None, max_length=128)
    append_artifact_ref: str | None = None


@router.post("")
async def create_task(request: Request, body: TaskCreateBody) -> dict:
    user = await get_current_user(request)
    tenant_id = db.user_tenant_id(user.id)
    if body.scope == "workspace":
        if body.workspace_id is None:
            raise HTTPException(status_code=400, detail="workspace_id required for workspace scope")
        if not user_may_access_workspace(user_id=user.id, workspace_id=body.workspace_id):
            raise HTTPException(status_code=403, detail="workspace not accessible")
    try:
        row = agent_tasks_store.create_task(
            tenant_id=tenant_id,
            created_by_user_id=user.id,
            scope=body.scope,
            goal=body.goal,
            task_type=body.task_type,
            workspace_id=body.workspace_id,
            parent_task_id=body.parent_task_id,
            conversation_id=body.conversation_id,
            status=body.status,  # type: ignore[arg-type]
            priority=body.priority,
            assigned_agent_id=body.assigned_agent_id,
            requirements=body.requirements,
            artifact_refs=body.artifact_refs,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return {"ok": True, "task": agent_tasks_store.row_to_public(row)}


@router.get("")
async def list_tasks(
    request: Request,
    scope: Literal["global", "workspace"] | None = None,
    workspace_id: uuid.UUID | None = None,
    parent_task_id: uuid.UUID | None = None,
    root_only: bool = False,
    status: str | None = None,
    limit: int = 50,
) -> dict:
    user = await get_current_user(request)
    tenant_id = db.user_tenant_id(user.id)
    if workspace_id is not None and not user_may_access_workspace(
        user_id=user.id, workspace_id=workspace_id
    ):
        raise HTTPException(status_code=403, detail="workspace not accessible")
    rows = agent_tasks_store.list_tasks(
        tenant_id=tenant_id,
        created_by_user_id=user.id,
        scope=scope,
        workspace_id=workspace_id,
        parent_task_id=parent_task_id,
        status=status,
        limit=limit,
    )
    if root_only:
        rows = [r for r in rows if r.get("parent_task_id") is None]
    return {"ok": True, "tasks": [agent_tasks_store.row_to_public(r) for r in rows]}


@router.get("/{task_id}")
async def get_task(request: Request, task_id: uuid.UUID) -> dict:
    user = await get_current_user(request)
    tenant_id = db.user_tenant_id(user.id)
    row = agent_tasks_store.get_task(task_id=task_id, tenant_id=tenant_id)
    if not row or not user_may_access_task_row(
        user_id=user.id, tenant_id=tenant_id, row=row
    ):
        raise HTTPException(status_code=404, detail="task not found")
    subs = agent_tasks_store.list_subtasks(
        tenant_id=tenant_id, parent_task_id=task_id, limit=100
    )
    arts = agent_artifacts_store.list_artifacts_for_task(
        tenant_id=tenant_id, task_id=task_id, limit=50
    )
    return {
        "ok": True,
        "task": agent_tasks_store.row_to_public(row),
        "subtasks": [agent_tasks_store.row_to_public(s) for s in subs],
        "artifacts": [agent_artifacts_store.row_to_public(a) for a in arts],
    }


@router.patch("/{task_id}")
async def patch_task(request: Request, task_id: uuid.UUID, body: TaskPatchBody) -> dict:
    user = await get_current_user(request)
    tenant_id = db.user_tenant_id(user.id)
    row = agent_tasks_store.get_task(task_id=task_id, tenant_id=tenant_id)
    if not row or not user_may_access_task_row(
        user_id=user.id, tenant_id=tenant_id, row=row
    ):
        raise HTTPException(status_code=404, detail="task not found")
    updated = agent_tasks_store.update_task(
        task_id=task_id,
        tenant_id=tenant_id,
        status=body.status,  # type: ignore[arg-type]
        goal=body.goal,
        priority=body.priority,
        assigned_agent_id=body.assigned_agent_id,
        append_artifact_ref=body.append_artifact_ref,
    )
    if not updated:
        raise HTTPException(status_code=500, detail="update failed")
    return {"ok": True, "task": agent_tasks_store.row_to_public(updated)}

File: apps/frontend/src/lib/tasksApi.ts
import { apiFetch } from "./api";
import type { AuthContextValue } from "../auth/AuthContext";

export type Task = {
  id: string;
  scope: "global" | "workspace";
  workspace_id?: string | null;
  parent_task_id?: string | null;
  goal: string;
  status: string;
  priority: string;
  assigned_agent_id?: string | null;
  artifact_refs?: string[];
  task_type?: string;
  updated_at?: string;
};

export type TaskDetail = {
  task: Task;
  subtasks: Task[];
  artifacts: { id: string; kind: string; summary: string }[];
};

export async function fetchTasks(
  auth: Pick<AuthContextValue, "accessToken" | "refresh">,
  params?: { scope?: string; workspace_id?: string; root_only?: boolean }
): Promise<Task[]> {
  const q = new URLSearchParams();
  if (params?.scope) q.set("scope", params.scope);
  if (params?.workspace_id) q.set("workspace_id", params.workspace_id);
  if (params?.root_only) q.set("root_only", "true");
  const path = `/v1/tasks${q.toString() ? `?${q}` : ""}`;
  const res = await apiFetch(path, auth);
  const data = (await res.json()) as { tasks?: Task[] };
  return data.tasks ?? [];
}

export async function fetchTask(
  auth: Pick<AuthContextValue, "accessToken" | "refresh">,
  taskId: string
): Promise<TaskDetail | null> {
  const res = await apiFetch(`/v1/tasks/${encodeURIComponent(taskId)}`, auth);
  if (!res.ok) return null;
  return (await res.json()) as TaskDetail;
}

export async function createTask(
  auth: Pick<AuthContextValue, "accessToken" | "refresh">,
  body: {
    scope: "global" | "workspace";
    goal: string;
    workspace_id?: string;
    parent_task_id?: string;
    conversation_id?: string;
  }
): Promise<Task> {
  const res = await apiFetch("/v1/tasks", auth, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  const data = (await res.json()) as { task: Task };
  return data.task;
}

export async function patchTask(
  auth: Pick<AuthContextValue, "accessToken" | "refresh">,
  taskId: string,
  body: { status?: string; goal?: string; priority?: string }
): Promise<Task> {
  const res = await apiFetch(`/v1/tasks/${encodeURIComponent(taskId)}`, auth, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  const data = (await res.json()) as { task: Task };
  return data.task;
}

export async function setConversationActiveTask(
  auth: Pick<AuthContextValue, "accessToken" | "refresh">,
  conversationId: string,
  activeTaskId: string | null
): Promise<void> {
  await apiFetch(
    `/v1/user/conversations/${encodeURIComponent(conversationId)}/active-task`,
    auth,
    {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ active_task_id: activeTaskId }),
    }
  );
}
